File: src/renderer/vulkan/renderpass/ShadowMap.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "ShadowMap.h"

#include "FrameGraphPass.h"
#include "MeshletCulling.h"
#include "ShadowConstants.h"

#include "renderer/PrepareBuckets.h"
#include "renderer/buffer/GPUBufferView.h"
#include "renderer/graph/FrameGraphBuilder.h"
#include "renderer/texture/GPUTextureProperties.h"
#include "renderer/vulkan/Backend.h"
#include "renderer/vulkan/CommandBuffer.h"
#include "renderer/vulkan/DescriptorSet.h"
#include "renderer/vulkan/FrameGraphResources.h"
#include "renderer/vulkan/GpuProfile.h"
#include "renderer/vulkan/Image.h"
#include "renderer/vulkan/Pipeline.h"
#include "renderer/vulkan/PipelineFactory.h"
#include "renderer/vulkan/RenderPassHelpers.h"
#include "renderer/vulkan/ShaderModules.h"
#include "renderer/vulkan/StorageBufferAllocator.h"

#include <array>
#include <fmt/format.h>

#include "renderer/shader/shadow/shadow_map_pass.share.hlsl"

namespace Reaper
{
namespace
{
    VkPipeline create_shadow_pipeline(VkDevice device, const ShaderModules& shader_modules,
                                      VkPipelineLayout pipeline_layout)

    {
        std::vector<VkPipelineShaderStageCreateInfo> shader_stages = {
            default_pipeline_shader_stage_create_info(VK_SHADER_STAGE_VERTEX_BIT, shader_modules.render_shadow_vs)};

        GraphicsPipelineProperties pipeline_properties = default_graphics_pipeline_properties();
        pipeline_properties.input_assembly.primitiveRestartEnable = VK_TRUE;
        pipeline_properties.depth_stencil.depthTestEnable = VK_TRUE;
        pipeline_properties.depth_stencil.depthWriteEnable = VK_TRUE;
        pipeline_properties.depth_stencil.depthCompareOp =
            ShadowUseReverseZ ? VK_COMPARE_OP_GREATER : VK_COMPARE_OP_LESS;
        pipeline_properties.pipeline_layout = pipeline_layout;
        pipeline_properties.pipeline_rendering.depthAttachmentFormat = PixelFormatToVulkan(ShadowMapFormat);

        std::vector<VkDynamicState> dynamic_states = {VK_DYNAMIC_STATE_VIEWPORT, VK_DYNAMIC_STATE_SCISSOR};

        return create_graphics_pipeline(device, shader_stages, pipeline_properties, dynamic_states);
    }
} // namespace

ShadowMapResources create_shadow_map_resources(VulkanBackend& backend, PipelineFactory& pipeline_factory)
{
    ShadowMapResources resources = {};

    std::vector<VkDescriptorSetLayoutBinding> descriptorSetLayoutBinding = {
        {0, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 1, VK_SHADER_STAGE_VERTEX_BIT, nullptr},
        {1, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 1, VK_SHADER_STAGE_VERTEX_BIT, nullptr},
    };

    resources.desc_set_layout = create_descriptor_set_layout(backend.device, descriptorSetLayoutBinding);
    resources.pipeline_layout = create_pipeline_layout(backend.device, std::span(&resources.desc_set_layout, 1));
    resources.pipeline_index = register_pipeline_creator(pipeline_factory,
                                                         PipelineCreator{
                                                             .pipeline_layout = resources.pipeline_layout,
                                                             .pipeline_creation_function = &create_shadow_pipeline,
                                                         });

    resources.descriptor_sets.resize(3); // FIXME

    allocate_descriptor_sets(backend.device, backend.global_descriptor_pool, resources.desc_set_layout,
                             resources.descriptor_sets);

    return resources;
}

void destroy_shadow_map_resources(VulkanBackend& backend, ShadowMapResources& resources)
{
    vkDestroyPipelineLayout(backend.device, resources.pipeline_layout, nullptr);
    vkDestroyDescriptorSetLayout(backend.device, resources.desc_set_layout, nullptr);
}

ShadowFrameGraphRecord create_shadow_map_pass_record(FrameGraph::Builder&                builder,
                                                     const CullMeshletsFrameGraphRecord& meshlet_pass,
                                                     const PreparedData&                 prepared)
{
    ShadowFrameGraphRecord record;
    record.pass_handle = builder.create_render_pass("Shadow");

    for (const ShadowPassData& shadow_pass : prepared.shadow_passes)
    {
        const GPUTextureProperties properties = default_texture_properties(
            shadow_pass.shadow_map_size.x, shadow_pass.shadow_map_size.y, ShadowMapFormat,
            GPUTextureUsage::DepthStencilAttachment | GPUTextureUsage::InputAttachment | GPUTextureUsage::Sampled);

        record.shadow_maps.push_back(builder.create_texture(
            record.pass_handle, "Shadow map", properties,
            GPUTextureAccess{VK_PIPELINE_STAGE_2_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_2_LATE_FRAGMENT_TESTS_BIT,
                             VK_ACCESS_2_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT, VK_IMAGE_LAYOUT_ATTACHMENT_OPTIMAL}));
    }

    record.meshlet_counters = builder.read_buffer(
        record.pass_handle, meshlet_pass.cull_triangles.meshlet_counters,
        GPUBufferAccess{VK_PIPELINE_STAGE_2_DRAW_INDIRECT_BIT, VK_ACCESS_2_INDIRECT_COMMAND_READ_BIT});

    record.meshlet_indirect_draw_commands = builder.read_buffer(
        record.pass_handle, meshlet_pass.cull_triangles.meshlet_indirect_draw_commands,
        GPUBufferAccess{VK_PIPELINE_STAGE_2_DRAW_INDIRECT_BIT, VK_ACCESS_2_INDIRECT_COMMAND_READ_BIT});

    record.meshlet_visible_index_buffer =
        builder.read_buffer(record.pass_handle, meshlet_pass.cull_triangles.meshlet_visible_index_buffer,
                            GPUBufferAccess{VK_PIPELINE_STAGE_2_INDEX_INPUT_BIT, VK_ACCESS_2_INDEX_READ_BIT});

    return record;
}

void update_shadow_map_resources(DescriptorWriteHelper& write_helper, StorageBufferAllocator& frame_storage_allocator,
                                 const PreparedData& prepared, ShadowMapResources& resources,
                                 GPUBuffer& vertex_position_buffer)
{
    REAPER_PROFILE_SCOPE_FUNC();

    if (prepared.shadow_instance_params.empty())
        return;

    const StorageBufferAlloc alloc = allocate_storage(frame_storage_allocator, prepared.shadow_instance_params.size()
                                                                                   * sizeof(ShadowMapInstanceParams));

    upload_storage_buffer(frame_storage_allocator, alloc, prepared.shadow_instance_params.data());

    for (const ShadowPassData& shadow_pass : prepared.shadow_passes)
    {
        if (shadow_pass.instance_count > 0)
        {
            Assert(shadow_pass.pass_index < resources.descriptor_sets.size());

            VkDescriptorSet descriptor_set = resources.descriptor_sets[shadow_pass.pass_index];

            const GPUBufferView instances_view = {
                .offset_bytes = alloc.offset_bytes + shadow_pass.instance_offset * sizeof(ShadowMapInstanceParams),
                .size_bytes = shadow_pass.instance_count * sizeof(ShadowMapInstanceParams),
            };

            write_helper.append(descriptor_set, 0, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, alloc.buffer,
                                instances_view.offset_bytes, instances_view.size_bytes);
            write_helper.append(descriptor_set, 1, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, vertex_position_buffer.handle);
        }
    }
}

void record_shadow_map_command_buffer(const FrameGraphHelper&       frame_graph_helper,
                                      const ShadowFrameGraphRecord& pass_record, CommandBuffer& cmdBuffer,
                                      const PipelineFactory& pipeline_factory, const PreparedData& prepared,
                                      ShadowMapResources& resources)
{
    REAPER_GPU_SCOPE(cmdBuffer, "Shadow");

    const FrameGraphBarrierScope framegraph_barrier_scope(cmdBuffer, frame_graph_helper, pass_record.pass_handle);

    const FrameGraphBuffer meshlet_counters = get_frame_graph_buffer(
        frame_graph_helper.resources, frame_graph_helper.frame_graph, pass_record.meshlet_counters);
    const FrameGraphBuffer meshlet_indirect_draw_commands = get_frame_graph_buffer(
        frame_graph_helper.resources, frame_graph_helper.frame_graph, pass_record.meshlet_indirect_draw_commands);
    const FrameGraphBuffer meshlet_visible_index_buffer = get_frame_graph_buffer(
        frame_graph_helper.resources, frame_graph_helper.frame_graph, pass_record.meshlet_visible_index_buffer);

    vkCmdBindPipeline(cmdBuffer.handle, VK_PIPELINE_BIND_POINT_GRAPHICS,
                      get_pipeline(pipeline_factory, resources.pipeline_index));

    for (const ShadowPassData& shadow_pass : prepared.shadow_passes)
    {
        if (shadow_pass.instance_count == 0)
            continue;

        VulkanDebugLabelCmdBufferScope s(cmdBuffer.handle,
                                         fmt::format("Shadow Pass {}", shadow_pass.pass_index).c_str());
        REAPER_PROFILE_SCOPE_GPU(cmdBuffer, "Shadow Pass");

        const FrameGraphTexture shadow_map =
            get_frame_graph_texture(frame_graph_helper.resources, frame_graph_helper.frame_graph,
                                    pass_record.shadow_maps[shadow_pass.pass_index]);
        const VkExtent2D output_extent = {shadow_pass.shadow_map_size.x, shadow_pass.shadow_map_size.y};
        const VkRect2D   pass_rect = default_vk_rect(output_extent);
        const VkViewport viewport = default_vk_viewport(pass_rect);

        Assert(output_extent.width == shadow_map.properties.width);
        Assert(output_extent.height == shadow_map.properties.height);

        vkCmdSetViewport(cmdBuffer.handle, 0, 1, &viewport);
        vkCmdSetScissor(cmdBuffer.handle, 0, 1, &pass_rect);

        VkRenderingAttachmentInfo depth_attachment =
            default_rendering_attachment_info(shadow_map.default_view_handle, shadow_map.image_layout);
        depth_attachment.loadOp = VK_ATTACHMENT_LOAD_OP_CLEAR;
        depth_attachment.clearValue = VkClearDepthStencil(ShadowUseReverseZ ? 0.f : 1.f, 0);

        const VkRenderingInfo rendering_info = default_rendering_info(pass_rect, nullptr, &depth_attachment);

        vkCmdBeginRendering(cmdBuffer.handle, &rendering_info);

        const MeshletDrawParams meshlet_draw = get_meshlet_draw_params(shadow_pass.pass_index);

        vkCmdBindIndexBuffer2(cmdBuffer.handle, meshlet_visible_index_buffer.handle, meshlet_draw.index_buffer_offset,
                              VK_WHOLE_SIZE, meshlet_draw.index_type);

        vkCmdBindDescriptorSets(cmdBuffer.handle, VK_PIPELINE_BIND_POINT_GRAPHICS, resources.pipeline_layout, 0, 1,
                                &resources.descriptor_sets[shadow_pass.pass_index], 0, nullptr);

        vkCmdDrawIndexedIndirectCount(cmdBuffer.handle, meshlet_indirect_draw_commands.handle,
                                      meshlet_draw.command_buffer_offset, meshlet_counters.handle,
                                      meshlet_draw.counter_buffer_offset, meshlet_draw.command_buffer_max_count,
                                      meshlet_indirect_draw_commands.properties.element_size_bytes);

        vkCmdEndRendering(cmdBuffer.handle);
    }
}
} // namespace Reaper

File: src/renderer/vulkan/renderpass/shadow_map.zig
// Port of src/renderer/vulkan/renderpass/ShadowMap.{h,cpp}
//
// One depth-only pass per shadow-casting light, each drawing that light's slice
// of the culled meshlets into its own depth target. The culling passes already
// run per shadow pass — `pass_index` indexes both the culling output and the
// descriptor set here — so this only has to bind and draw.

const std = @import("std");
const vk = @import("vulkan");

const constants = @import("constants.zig");
const descriptor_set = @import("../descriptor_set.zig");
const fg = @import("../../graph/frame_graph.zig");
const frame_graph_pass = @import("frame_graph_pass.zig");
const gpu_buffer = @import("../../buffer/gpu_buffer.zig");
const gpu_texture_properties = @import("../../texture/gpu_texture_properties.zig");
const meshlet_culling = @import("meshlet_culling.zig");
const pipeline_module = @import("../pipeline.zig");
const pipeline_factory_module = @import("../pipeline_factory.zig");
const render_pass_helpers = @import("../render_pass_helpers.zig");
const shader_modules = @import("../shader_modules.zig");
const storage_buffer = @import("../storage_buffer.zig");

const hlsl_shadow = @import("../../hlsl/shadow/shadow_map_pass.zig");

const Builder = @import("../../graph/builder.zig").Builder;
const DescriptorWriteHelper = descriptor_set.DescriptorWriteHelper;
const FrameGraphResources = @import("../framegraph_resources.zig").FrameGraphResources;
const PipelineFactory = pipeline_factory_module.PipelineFactory;
const prepare_buckets = @import("../../prepare_buckets.zig");

/// Port of ShadowConstants.h.
pub const shadow_use_reverse_z = prepare_buckets.shadow_use_reverse_z;
pub const shadow_map_format: vk.Format = .d16_unorm;

/// FIXME The C++ allocates a fixed three sets up front rather than sizing them
/// to the scene.
const descriptor_set_count: u32 = 3;

// --------------------------------------------------------------------------
// Descriptor bindings
// --------------------------------------------------------------------------

const SetZero = struct {
    const instance_params = 0;
    const buffer_position_ms = 1;

    const bindings = [_]descriptor_set.DescriptorBinding{
        .{ .slot = 0, .count = 1, .type = .storage_buffer, .stage_mask = .{ .vertex_bit = true } },
        .{ .slot = 1, .count = 1, .type = .storage_buffer, .stage_mask = .{ .vertex_bit = true } },
    };
};

// --------------------------------------------------------------------------
// Pipeline
// --------------------------------------------------------------------------

fn createShadowPipeline(
    vkd: *const vk.DeviceWrapper,
    device: vk.Device,
    pipeline_layout: vk.PipelineLayout,
) anyerror!vk.Pipeline {
    const module_create_info_vert = pipeline_module.shaderModuleCreateInfo(
        shader_modules.get("shadow/render_shadow.vert.spv"),
    );

    // Depth only: there is no fragment stage.
    const shader_stages = [_]vk.PipelineShaderStageCreateInfo{
        pipeline_module.defaultPipelineShaderStageCreateInfo(.{ .vertex_bit = true }, &module_create_info_vert, null),
    };

    var properties = pipeline_module.defaultGraphicsPipelineProperties(null);

    // Meshlet index buffers use a primitive restart value between triangles.
    properties.input_assembly.primitive_restart_enable = .true;

    properties.depth_stencil.depth_test_enable = .true;
    properties.depth_stencil.depth_write_enable = .true;
    properties.depth_stencil.depth_compare_op = if (shadow_use_reverse_z) .greater else .less;
    properties.pipeline_layout = pipeline_layout;
    properties.pipeline_rendering.depth_attachment_format = shadow_map_format;

    const dynamic_states = [_]vk.DynamicState{ .viewport, .scissor };

    return pipeline_module.createGraphicsPipeline(vkd, device, &shader_stages, &properties, &dynamic_states);
}

// --------------------------------------------------------------------------
// Resources
// --------------------------------------------------------------------------

pub const ShadowMapResources = struct {
    pipeline_index: u32,
    pipeline_layout: vk.PipelineLayout,
    desc_set_layout: vk.DescriptorSetLayout,

    descriptor_sets: [descriptor_set_count]vk.DescriptorSet,

    pub fn init(
        vkd: anytype,
        device: vk.Device,
        descriptor_pool: vk.DescriptorPool,
        pipeline_factory: *PipelineFactory,
    ) !ShadowMapResources {
        var bindings: [SetZero.bindings.len]vk.DescriptorSetLayoutBinding = undefined;
        descriptor_set.fillLayoutBindings(&bindings, &SetZero.bindings);

        const binding_flags = [_]vk.DescriptorBindingFlags{.{}} ** SetZero.bindings.len;

        const desc_set_layout = try pipeline_module.createDescriptorSetLayout(vkd, device, &bindings, &binding_flags);
        errdefer vkd.destroyDescriptorSetLayout(device, desc_set_layout, null);

        const pipeline_layout = try pipeline_module.createPipelineLayout(vkd, device, &.{desc_set_layout}, &.{});
        errdefer vkd.destroyPipelineLayout(device, pipeline_layout, null);

        const pipeline_index = try pipeline_factory.registerPipelineCreator(.{
            .pipeline_layout = pipeline_layout,
            .pipeline_creation_function = &createShadowPipeline,
        });

        const set_layouts = [_]vk.DescriptorSetLayout{desc_set_layout} ** descriptor_set_count;

        var descriptor_sets: [descriptor_set_count]vk.DescriptorSet = undefined;
        try pipeline_module.allocateDescriptorSets(vkd, device, descriptor_pool, &set_layouts, &descriptor_sets);

        return .{
            .pipeline_index = pipeline_index,
            .pipeline_layout = pipeline_layout,
            .desc_set_layout = desc_set_layout,
            .descriptor_sets = descriptor_sets,
        };
    }

    pub fn deinit(self: *ShadowMapResources, vkd: anytype, device: vk.Device) void {
        vkd.destroyPipelineLayout(device, self.pipeline_layout, null);
        vkd.destroyDescriptorSetLayout(device, self.desc_set_layout, null);
    }
};

// --------------------------------------------------------------------------
// Frame graph record
// --------------------------------------------------------------------------

pub const ShadowFrameGraphRecord = struct {
    pass_handle: fg.RenderPassHandle,
    /// One per entry of `prepared.shadow_passes`, in the same order, so
    /// `pass_index` indexes this directly.
    shadow_maps: []const fg.ResourceUsageHandle,
    meshlet_counters: fg.ResourceUsageHandle,
    meshlet_indirect_draw_commands: fg.ResourceUsageHandle,
    meshlet_visible_index_buffer: fg.ResourceUsageHandle,
};

pub fn createFrameGraphRecord(
    builder: *Builder,
    allocator: std.mem.Allocator,
    meshlet_pass: meshlet_culling.CullMeshletsFrameGraphRecord,
    prepared: *const prepare_buckets.PreparedData,
) !ShadowFrameGraphRecord {
    const pass_handle = try builder.createRenderPass("Shadow", false);

    const shadow_maps = try allocator.alloc(fg.ResourceUsageHandle, prepared.shadow_passes.items.len);

    for (prepared.shadow_passes.items, shadow_maps) |shadow_pass, *out| {
        const properties = gpu_texture_properties.defaultTextureProperties(
            shadow_pass.shadow_map_size[0],
            shadow_pass.shadow_map_size[1],
            shadow_map_format,
            .{ .depth_stencil_attachment = true, .input_attachment = true, .sampled = true },
        );

        out.* = try builder.createTexture(
            pass_handle,
            "Shadow map",
            properties,
            .{
                .stage_mask = .{ .early_fragment_tests_bit = true, .late_fragment_tests_bit = true },
                .access_mask = .{ .depth_stencil_attachment_write_bit = true },
                .image_layout = .attachment_optimal,
            },
            &.{},
        );
    }

    const meshlet_counters = try builder.readBuffer(
        pass_handle,
        meshlet_pass.cull_triangles.meshlet_counters,
        .{ .stage_mask = .{ .draw_indirect_bit = true }, .access_mask = .{ .indirect_command_read_bit = true } },
        &.{},
    );

    const meshlet_indirect_draw_commands = try builder.readBuffer(
        pass_handle,
        meshlet_pass.cull_triangles.meshlet_indirect_draw_commands,
        .{ .stage_mask = .{ .draw_indirect_bit = true }, .access_mask = .{ .indirect_command_read_bit = true } },
        &.{},
    );

    const meshlet_visible_index_buffer = try builder.readBuffer(
        pass_handle,
        meshlet_pass.cull_triangles.meshlet_visible_index_buffer,
        .{ .stage_mask = .{ .index_input_bit = true }, .access_mask = .{ .index_read_bit = true } },
        &.{},
    );

    return .{
        .pass_handle = pass_handle,
        .shadow_maps = shadow_maps,
        .meshlet_counters = meshlet_counters,
        .meshlet_indirect_draw_commands = meshlet_indirect_draw_commands,
        .meshlet_visible_index_buffer = meshlet_visible_index_buffer,
    };
}

// --------------------------------------------------------------------------
// Descriptor updates
// --------------------------------------------------------------------------

pub fn updateDescriptorSets(
    write_helper: *DescriptorWriteHelper,
    frame_storage_allocator: *storage_buffer.StorageBufferAllocator,
    prepared: *const prepare_buckets.PreparedData,
    resources: *const ShadowMapResources,
    vertex_position_buffer: vk.Buffer,
) void {
    if (prepared.shadow_instance_params.items.len == 0) return;

    const alloc = frame_storage_allocator.allocateAndUpload(
        hlsl_shadow.ShadowMapInstanceParams,
        prepared.shadow_instance_params.items,
    );

    for (prepared.shadow_passes.items) |shadow_pass| {
        if (shadow_pass.instance_count == 0) continue;

        std.debug.assert(shadow_pass.pass_index < descriptor_set_count);

        const set = resources.descriptor_sets[shadow_pass.pass_index];

        const instances_view = gpu_buffer.GPUBufferView{
            .offset_bytes = alloc.offset_bytes +
                shadow_pass.instance_offset * @sizeOf(hlsl_shadow.ShadowMapInstanceParams),
            .size_bytes = shadow_pass.instance_count * @sizeOf(hlsl_shadow.ShadowMapInstanceParams),
        };

        write_helper.appendBuffer(
            set,
            SetZero.instance_params,
            .storage_buffer,
            alloc.buffer,
            instances_view.offset_bytes,
            instances_view.size_bytes,
        );
        write_helper.appendBuffer(
            set,
            SetZero.buffer_position_ms,
            .storage_buffer,
            vertex_position_buffer,
            0,
            vk.WHOLE_SIZE,
        );
    }
}

// --------------------------------------------------------------------------
// Recording
// --------------------------------------------------------------------------

pub fn recordCommandBuffer(
    vkd: anytype,
    cmd_buffer: vk.CommandBuffer,
    helper: frame_graph_pass.FrameGraphHelper,
    pipeline_factory: *const PipelineFactory,
    record: ShadowFrameGraphRecord,
    prepared: *const prepare_buckets.PreparedData,
    resources: *const ShadowMapResources,
) void {
    frame_graph_pass.beginBarrierScope(vkd, cmd_buffer, helper, record.pass_handle);
    defer frame_graph_pass.endBarrierScope(vkd, cmd_buffer, helper, record.pass_handle);

    const meshlet_counters = helper.resources.getBuffer(helper.frame_graph, record.meshlet_counters);
    const indirect_draw_commands = helper.resources.getBuffer(helper.frame_graph, record.meshlet_indirect_draw_commands);
    const visible_index_buffer = helper.resources.getBuffer(helper.frame_graph, record.meshlet_visible_index_buffer);

    vkd.cmdBindPipeline(cmd_buffer, .graphics, pipeline_factory.getPipeline(resources.pipeline_index));

    for (prepared.shadow_passes.items) |shadow_pass| {
        if (shadow_pass.instance_count == 0) continue;

        const shadow_map = helper.resources.getTexture(helper.frame_graph, record.shadow_maps[shadow_pass.pass_index]);

        const extent = vk.Extent2D{
            .width = shadow_pass.shadow_map_size[0],
            .height = shadow_pass.shadow_map_size[1],
        };

        std.debug.assert(extent.width == shadow_map.properties.width);
        std.debug.assert(extent.height == shadow_map.properties.height);

        const pass_rect = render_pass_helpers.defaultRect(extent);
        const viewports = [_]vk.Viewport{render_pass_helpers.defaultViewport(pass_rect)};
        const scissors = [_]vk.Rect2D{pass_rect};

        vkd.cmdSetViewport(cmd_buffer, 0, &viewports);
        vkd.cmdSetScissor(cmd_buffer, 0, &scissors);

        var depth_attachment = pipeline_module.defaultRenderingAttachmentInfo(
            shadow_map.default_view_handle,
            shadow_map.image_layout,
        );
        depth_attachment.load_op = .clear;
        depth_attachment.clear_value = .{
            .depth_stencil = .{ .depth = if (shadow_use_reverse_z) 0.0 else 1.0, .stencil = 0 },
        };

        const rendering_info = pipeline_module.defaultRenderingInfo(pass_rect, &.{}, &depth_attachment);

        vkd.cmdBeginRendering(cmd_buffer, &rendering_info);

        const meshlet_draw = meshlet_culling.getMeshletDrawParams(shadow_pass.pass_index);

        vkd.cmdBindIndexBuffer2(
            cmd_buffer,
            visible_index_buffer.handle,
            meshlet_draw.index_buffer_offset,
            vk.WHOLE_SIZE,
            meshlet_draw.index_type,
        );

        const pass_descriptors = [_]vk.DescriptorSet{resources.descriptor_sets[shadow_pass.pass_index]};
        vkd.cmdBindDescriptorSets(cmd_buffer, .graphics, resources.pipeline_layout, 0, &pass_descriptors, &.{});

        vkd.cmdDrawIndexedIndirectCount(
            cmd_buffer,
            indirect_draw_commands.handle,
            meshlet_draw.command_buffer_offset,
            meshlet_counters.handle,
            meshlet_draw.counter_buffer_offset,
            meshlet_draw.command_buffer_max_count,
            indirect_draw_commands.properties.element_size_bytes,
        );

        vkd.cmdEndRendering(cmd_buffer);
    }
}

// --------------------------------------------------------------------------
// Tests
// --------------------------------------------------------------------------

const testing = std.testing;

test "the shadow pipeline's depth format matches the frame graph texture" {
    // The pipeline declares this up front and the frame graph creates the
    // texture separately, so a mismatch only shows up as a validation error at
    // draw time.
    try testing.expectEqual(vk.Format.d16_unorm, shadow_map_format);
}

test "reverse z picks the matching compare op and clear value" {
    const compare_op: vk.CompareOp = if (shadow_use_reverse_z) .greater else .less;
    const clear_depth: f32 = if (shadow_use_reverse_z) 0.0 else 1.0;

    if (shadow_use_reverse_z) {
        try testing.expectEqual(vk.CompareOp.greater, compare_op);
        try testing.expectEqual(@as(f32, 0.0), clear_depth);
    } else {
        try testing.expectEqual(vk.CompareOp.less, compare_op);
        try testing.expectEqual(@as(f32, 1.0), clear_depth);
    }
}

test "shadow passes never outnumber the descriptor sets or culling passes" {
    // `pass_index` indexes both `descriptor_sets` here and the culling output,
    // and prepare_buckets reserves one culling pass for the main view.
    try testing.expect(descriptor_set_count <= meshlet_culling.max_meshlet_culling_pass_count - 1);
}

test "descriptor set zero bindings are dense and correctly ordered" {
    for (SetZero.bindings, 0..) |binding, i| {
        try testing.expectEqual(@as(u32, @intCast(i)), binding.slot);
    }

    try testing.expectEqual(vk.DescriptorType.storage_buffer, SetZero.bindings[SetZero.instance_params].type);
    try testing.expectEqual(vk.DescriptorType.storage_buffer, SetZero.bindings[SetZero.buffer_position_ms].type);
}
